### ISSUE TITLE
Using [default-run] and `cargo run` doesn't build all binaries

### ISSUE THREAD

**Issue by jonathandturner** [rust-lang/cargo#7387]
**Problem**

**Steps**

1. Use the [default-run] feature
2. Run the project with `cargo run`
3. Notice that not all binaries were built

**Possible Solution(s)**

**Notes**

Apologies if this is a duplicate, I couldn't find it.

Something we noticed with nushell is that to build Nu and all the plugins, we manually have to do `cargo build && cargo run` instead of just `cargo run`. This is because `cargo run` by itself only builds what's needed for the binary pointed to by `default-run`. Is there a way to configure the Cargo.toml to always build all the binaries (since in Nu they work together), or to build the binaries and run the one pointed to by `default-run`?

Output of `cargo version`:

cargo 1.39.0-nightly (fe0e5a48b 2019-09-04)

**Comment by lukaslueg** [comment]
I'm not sure if building all executables was desireable because everyone would have to pay to build all the binaries all the time, even if only the main binary is going to get executed.

**Comment by jonathandturner** [comment]
Agreed that not everyone would want that behavior. Perhaps the ask here is to be able to configure it for those of us who need it.

**Comment by djrodgerspryor** [comment]
I'd love to be able to specify that one binary depends on another in `Cargo.toml` so that `run` could build exactly what's actually needed.

My use-case involves binaries which need to talk to each other at runtime.

**Comment by mrcnski** [comment]
For the single-binary case, my mental model for `cargo run` has been that it's basically `cargo build` followed by an invocation of the binary. So this behavior for multiple binaries is counterintuitive.

> I'd love to be able to specify that one binary depends on another in Cargo.toml so that run could build exactly what's actually needed.

Perhaps one could get around this with [artifact dependencies](https://doc.rust-lang.org/cargo/reference/unstable.html#artifact-dependencies), though it's unfortunately an unstable feature.

**Comment by s0me0ne-unkn0wn** [comment]
I do agree, it's totally counterintuitive. I haven't even ever suspected it could behave differently, nor checked it, it seemed to be obvious. And it's a really bad UX for users, we should fix it somehow.

**Comment by eskimor** [comment]
Would a PR making this behavior configurable be accepted @lukaslueg ? We are already discussing ugly workarounds of the binary being built extracting the other binaries from itself ... I would much rather invest the time improving cargo.

If so, we would do our best to minimize review workload on your end: Doing proper research on the best/most idiomatic way of implementing this, creating docs, ... - pointers would of course be very much appreciated as well.

**Comment by lukaslueg** [comment]
@eskimor I'm not an authority on this. Just passing by.
